### ISSUE TITLE
ENH: propagate_dft should accept an optional mask

### ISSUE THREAD

**Issue by andykee** [andykee/lentil#44]
### Proposed new feature or change

The current behavior of `propagate_dft()` checks for overlap between the propagation result and the full-sized output canvas before performing a propagation. It would be beneficial to allow the specification of a mask to further refine the areas of interest within the output canvas. 

To accomplish this, the `propagate._overlap()` method could be updated to return one of two things:

1. An empty set: `()`
2. Bounding box coordinates

With this information, we can compute the appropriate `shape` and `offset` for the DFT calculation.

### Backward compatibility
This is a new feature and does not affect backward compatibility

**Comment by andykee** [comment]
The snippet below holds some promise, although there is more work to be done. Basically, it calculates the shape and new center of a masked, shifted field.

```python
def _calc_shape_shift(output_shape, field_shape, field_shift, mask=None):
    """Compute propagation shape and shift
    
    Parameters
    ----------
    output_shape : (2,) array_like
        Propagation output shape
    field_shape : (2,) array_like
        Field shape
    field_shift : (2,) array_like
        Field shift (from center of output array)
    mask : array_like, optional
        Binary mask used to identify invalid portions of the output shape. 
        If not supplied (default), the propagation shape and shift are 
        computed assuming the entire output shape is available for use.
        If mask is provided, it must have the same shape as output_shape 
        and will be cast to a bool before use.

    Returns
    -------
    shape : tuple
        Propagation shape. If shifted/masked field does not overlap with
        output array, shape is ()
    shift : (2,) tuple
        Field shift updated to account for additional masking

    """
    output_shape = tuple(int(n) for n in output_shape)
    field_shape = tuple(int(n) for n in field_shape)
    field_shift = tuple(int(n) for n in field_shift)
    field_shifted_ul = (output_shape[0]//2 - field_shape[0]//2 + field_shift[0],
                        output_shape[1]//2 - field_shape[1]//2 + field_shift[1])

    if mask is None:
        # no mask, just evaluate whether any part of the shifted field
        # falls within the output shape 
        shape, shift = field_shape, field_shift
        if field_shifted_ul[0] > output_shape[0]:
            shape = ()
        if field_shifted_ul[0] + field_shape[0] < 0:
            shape = ()
        if field_shifted_ul[1] > output_shape[1]:
            shape = ()
        if field_shifted_ul[1] + field_shape[1] < 0:
            shape = ()

    else:
        mask = np.asarray(mask, dtype=bool)
        if mask.shape != output_shape:
            pass
        rslice = slice(field_shifted_ul[0], field_shifted_ul[0]+field_shape[0])
        cslice = slice(field_shifted_ul[1], field_shifted_ul[1]+field_shape[1])
        mask_slice = mask[rslice, cslice]

        if not np.any(mask_slice):
            # field is completely masked out
            shape, shift = (), field_shift
        else:
            # compute unmasked portion of field and update shift accordingly
            (rmin, rmax, cmin, cmax) = lentil.boundary(mask_slice)
            print('_calc_shape_shift mask boundary:', rmin, rmax, cmin, cmax)
            masked_shape = (rmax-rmin+1, cmax-cmin+1)
            print('_calc_shape_shift masked_shape:', masked_shape)
            masked_center = (rmin + masked_shape[0]/2, cmin + masked_shape[1]/2)
            field_center = (field_shape[0]/2, field_shape[1]/2)
            masked_shift = (masked_center[0] - field_center[0],
                            masked_center[1] - field_center[1])
            #shape, shift = masked_shape, field_shift


            # update provided shift with additional shift due to masking 
            # operation
            shape = masked_shape
            shift = (field_shift[0] + masked_shift[0], 
                     field_shift[1] + masked_shift[1])
        
    return tuple(shape), tuple(shift)
```

In theory, this could be inserted into `propagate_dft()` like so:

```python
        prop_shape_out, fix_shift = _calc_shape_shift(shape_out, 
                                                      prop_shape_out, 
                                                      fix_shift, mask)
        if prop_shape_out:
            alpha = lentil.helper.dft_alpha(dx=dx, du=du, z=z,
                                            wave=wavefront.wavelength,
                                            oversample=oversample)
            data = lentil.fourier.dft2(f=field.data, alpha=alpha,
                                       shape=prop_shape_out,
                                       shift=subpx_shift,
                                       offset=field.offset, unitary=True)
            out.data.append(Field(data=data, pixelscale=du/oversample,
                                  offset=fix_shift))
```

This doesn't quite get us all the way there though. We also need to modify `fourier.dft2()` to accept shifts in both the input and output. Thankfully, this was already prototyped for a special project earlier this year:

```python
def dft2(f, alpha, npix=None, shift=(0, 0), input_center=(0,0), output_center=(0,0), unitary=True, out=None):
    r"""Compute the 2-dimensional discrete Fourier Transform.

    The DFT is defined in one dimension as

    .. math::

        A_k = \sum_{m=0}^{n-1} a_n \exp \left(-2\pi i\frac{mk}{n}\right) \hspace{3em} k=0, \dots, n-1

    This function allows independent control over input shape, output shape,
    and output sampling by implementing the matrix triple product algorithm
    described in [1].

    Parameters
    ----------
    f : array_like
        2D array to Fourier Transform
    alpha : float or array_like
        Output plane sampling interval (frequency). If :attr:`alpha` is an
        array, ``alpha[1]`` represents row-wise sampling and ``alpha[2]``
        represents column-wise sampling. If :attr:`alpha` is a scalar,
        ``alpha[1] = alpha[2] = alpha`` gives uniform sampling across the rows
        and columns of the output plane.
    npix : int or array_like, optional
        Size of the output array :attr:`F`. If :attr:`npix` is an array,
        ``F.shape = (npix[1], npix[2])``. If :attr:`npi`` is a scalar,
        ``F.shape = (npix, npix)``. Default is ``f.shape``.
    shift : array_like, optional
        Number of pixels in (r,c) to shift the DC pixel in the output plane
        with the origin centrally located in the plane. Default is ``(0,0)``.
    offset : array_like, optional
        Number of pixels in (r,c) that the input plane is shifted relative to
        the origin. Default is ``(0,0)``.
    unitary : bool, optional
        Normalization flag. If ``True``, a normalization is performed on the
        output such that the DFT operation is unitary and energy is conserved
        through the Fourier transform operation (Parseval's theorem). In this
        way, the energy in in a limited-area DFT is a fraction of the total
        energy corresponding to the limited area. Default is ``True``.
    out : ndarray or None
        A location into which the result is stored. If provided, it must have
        shape = npix and dtype = np.complex. If not provided or None, a
        freshly-allocated array is returned.

    Returns
    -------
    F : complex ndarray

    Notes
    -----
    * Setting ``alpha = 1/f.shape`` and ``npix = f.shape`` is equivalent to

      ::

          F = np.fft.ifftshift(np.fft.fft2(np.fft.fftshift(f)))

    * ``dft2()`` is designed to place the DC pixel in the same location as a
      well formed call to any standard FFT for both even and odd sized input
      arrays. The DC pixel is located at ``np.floor(npix/2) + 1``, which is
      consistent with calls to Numpy's FFT method where the input and output
      are correctly shifted:
      ``np.fft.ifftshift(np.fft.fft2(np.fft.fftshift(f)))``.

    * If the y-axis shift behavior is not what you are expecting, you most
      likely have your plotting axes flipped (matplotlib's default behavior is
      to place [0,0] in the upper left corner of the axes). This may be resolved
      by either flipping the sign of the y component of ``shift`` or by passing
      ``origin = 'lower'`` to ``imshow()``.

    References
    ----------
    [1] Soummer, et. al. Fast computation of Lyot-style coronagraph propagation (2007)

    """
    alpha_row, alpha_col = np.broadcast_to(np.asarray(alpha), (2,))

    f = np.asarray(f)
    m, n = f.shape

    if npix is None:
        npix = [m, n]
    M, N = np.broadcast_to(np.asarray(npix, dtype=int), (2,))

    shift_row, shift_col = np.broadcast_to(np.asarray(shift, dtype=int), (2,))
    input_center_row, input_center_col = np.broadcast_to(np.asarray(input_center, dtype=int), (2,))
    output_center_row, output_center_col = np.broadcast_to(np.asarray(output_center, dtype=int), (2,))

    if out is not None:
        if not np.can_cast(complex, out.dtype):
            raise TypeError(f"Cannot cast complex output to dtype('{out.dtype}')")

    E1, E2 = _dft2_matrices(m, n, M, N, alpha_row, alpha_col, shift_row, shift_col,
                            input_center_row, input_center_col, output_center_row, output_center_col)
    F = np.dot(E1.dot(f), E2, out=out)

    # now calculate the answer, without reallocating memory
    if unitary:
        np.multiply(F, np.sqrt(np.abs(alpha_row * alpha_col)), out=F)

    return F


@functools.lru_cache(maxsize=32)
def _dft2_matrices(m, n, M, N, alphar, alphac, shiftr, shiftc, inr, inc, outr, outc):
    R, S, U, V = _dft2_coords(m, n, M, N)
    E1 = np.exp(-2.0 * 1j * np.pi * alphar * np.outer(R-shiftr+inr, U-shiftr+outr)).T
    E2 = np.exp(-2.0 * 1j * np.pi * alphac * np.outer(S-shiftc+inc, V-shiftc+outc))
    return E1, E2


@functools.lru_cache(maxsize=32)
def _dft2_coords(m, n, M, N):
    # R and S are (r,c) coordinates in the (m x n) input plane f
    # V and U are (r,c) coordinates in the (M x N) output plane F

    R = np.arange(m) - np.floor(m/2.0)
    S = np.arange(n) - np.floor(n/2.0)
    U = np.arange(M) - np.floor(M/2.0)
    V = np.arange(N) - np.floor(N/2.0)

    return R, S, U, V
```

**Comment by andykee** [comment]
It's also worth considering whether we can reuse the helper functions [boundary_slice()](https://lentil.readthedocs.io/en/latest/generated/lentil.helper.boundary_slice.html#lentil.helper.boundary_slice) and [slice_offset()](https://lentil.readthedocs.io/en/latest/generated/lentil.helper.slice_offset.html#lentil.helper.slice_offset) inside `_calc_shape_shift()`.

**Comment by andykee** [comment]
Resolved in cad2faa
